pkg/actuators/machineset/controller_test.go

```diff
@@ -0,0 +1,359 @@
+			if err != nil {
+				return nil
+			}
+			annotations := m.GetAnnotations()
+			if annotations != nil {
+				return annotations
+			}
+			// Return an empty map to distinguish between empty annotations and errors
+			return make(map[string]string)
+		}, timeout).Should(Equal(rtc.expectedAnnotations))
+
+		// Check which event types were sent
+		Eventually(fakeRecorder.Events, timeout).Should(HaveLen(len(rtc.expectedEvents)))
+		receivedEvents := []string{}
+		eventMatchers := []gtypes.GomegaMatcher{}
+		for _, eachEvent := range rtc.expectedEvents {
+			receivedEvents = append(receivedEvents, <-fakeRecorder.Events)
+			eventMatchers = append(eventMatchers, ContainSubstring(fmt.Sprintf(" %s ", eachEvent)))
+		}
+		Expect(receivedEvents).To(ConsistOf(eventMatchers))
+	},
+		Entry("with no profile set", reconcileTestCase{
+			profile:             "",
+			existingAnnotations: make(map[string]string),
+			expectedAnnotations: make(map[string]string),
+			expectedEvents:      []string{"FailedUpdate"},
+		}),
+		Entry("with a bx2d-4x16", reconcileTestCase{
+			profile:             "bx2d-4x16",
+			existingAnnotations: make(map[string]string),
+			expectedAnnotations: map[string]string{
+				profileKey: "bx2d-4x16",
+			},
+			expectedEvents: []string{},
+		}),
+		Entry("with a bx2-2x8", reconcileTestCase{
+			profile:             "bx2-2x8",
+			existingAnnotations: make(map[string]string),
+			expectedAnnotations: map[string]string{
+				profileKey: "bx2-2x8",
+			},
+			expectedEvents: []string{},
+		}),
+		Entry("with existing annotations", reconcileTestCase{
+			profile: "bx2-2x8",
+			existingAnnotations: map[string]string{
+				"existing": "annotation",
+				"annother": "existingAnnotation",
+			},
+			expectedAnnotations: map[string]string{
+				"existing": "annotation",
+				"annother": "existingAnnotation",
+				profileKey: "bx2-2x8",
+			},
+			expectedEvents: []string{},
+		}),
+		Entry("with an invalid instance profile", reconcileTestCase{
+			profile: "invalid",
+			existingAnnotations: map[string]string{
+				"existing": "annotation",
+				"annother": "existingAnnotation",
+			},
+			expectedAnnotations: map[string]string{
+				"existing": "annotation",
+				"annother": "existingAnnotation",
+			},
+			expectedEvents: []string{"FailedUpdate"},
+		}),
+	)
+})
+
+func deleteMachineSets(c client.Client, namespaceName string) error {
+	machineSets := &machinev1.MachineSetList{}
+	err := c.List(ctx, machineSets, client.InNamespace(namespaceName))
+	if err != nil {
+		return err
+	}
+
+	for _, ms := range machineSets.Items {
+		err := c.Delete(ctx, &ms)
+		if err != nil {
+			return err
+		}
+	}
+
+	Eventually(func() error {
+		machineSets := &machinev1.MachineSetList{}
+		err := c.List(ctx, machineSets)
+		if err != nil {
+			return err
+		}
+		if len(machineSets.Items) > 0 {
+			return fmt.Errorf("machineSets not deleted")
+		}
+		return nil
+	}, timeout).Should(Succeed())
+
+	return nil
+}
+
+func TestReconcile(t *testing.T) {
```

Comment:
This set of tests seems to do exactly the same thing as the one above, except one calls the reconcile manually and the other relies on the controllers running in env test 🤔

Reply:
I was just following GCP and AWS. I went ahead and removed one set of the tests.